### PR TITLE
Add open graph meta tags to view layout

### LINKED ISSUE #4957: Provide meta og tags for various pages
**Is your feature request related to a problem? Please describe.**
When I get links from people, about something on OBS on IRC, I get message from bot, saying what exactly that link links to. When I'm using OBS links on any other platform however, that functionality would require me to write a bot for every chat app I use. Most of them however support embeding links with meta og tags.

**Describe the solution you'd like**
I would like OBS to provide useful information to see what is inside the link before I click it.

**Additional context**
Github provides sizable feedback from link:
example

### PR DESCRIPTION
**Pull request by SDBowen**
Added default Open Graph meta tags to the views layout. On most pages the tags will look like this:

Screenshot from 2019-03-18 01-53-29

On Project, Package, Request, or User show pages the meta tags will have more specific detail.

Screenshot from 2019-03-18 01-49-00

The OBS image I chose is probably not the best (svg with transparent background), as it will not look great on a white background. Is there a better image that I can replace it with?

Fixes #4957

### CODE REVIEW

File: src/api/app/views/layouts/webui2/webui.html.haml [hunk 1]
```diff
@@ -9,6 +9,13 @@
     %title
       = "#{@pagetitle} - " if @pagetitle
       = home_title
+    %meta{ property: "og:title", content: "#{meta_title}"}/
```
Comment (coolo) [inline src/api/app/views/layouts/webui2/webui.html.haml]: content: meta_title should do - no need to interpolate.

File: src/api/app/views/webui/request/show.html.erb [hunk 2]
```diff
@@ -2,6 +2,10 @@
 <% @layouttype = 'custom' %>
 <% @crumb_list = ["Request #{@bs_request.number}"] %>
 
+<% content_for(:meta_title, "#{@pagetitle} (#{@bs_request.state})") %>
+<% content_for(:meta_description, @bs_request.description) %>
+<% content_for(:meta_image, gravatar_icon((User.find_by_login(@bs_request.creator).email), 500)) %>
```
Comment (coolo) [inline src/api/app/views/webui/request/show.html.erb]: 500px are a lot - is that normal for meta_image?

File: src/api/app/views/layouts/webui2/webui.html.haml [hunk 3]
```diff
@@ -9,6 +9,13 @@
...
+    %meta{ property: "og:type", content: 'website'}/
+    %meta{ property: "og:url", content: request.original_url }/
+    %meta{ property: "og:image", content: "#{meta_image}"}/
```
Comment (dmarcoux) [inline src/api/app/views/layouts/webui2/webui.html.haml]: Same here, no need to interpolate.

File: src/api/app/views/layouts/webui2/webui.html.haml [hunk 4]
```diff
@@ -9,6 +9,13 @@
...
+    - if meta_description
+      %meta{ property: "og:description", content: "#{meta_description}"}/
```
Comment (dmarcoux) [inline src/api/app/views/layouts/webui2/webui.html.haml]: Again, no interpolation needed.

File: src/api/app/views/webui/user/show.html.haml [hunk 5]
```diff
@@ -1,6 +1,10 @@
 - @pagetitle = "Home of #{@displayed_user}"
 - @layouttype = 'custom'
 - @crumb_list = [@pagetitle]
+
+- content_for(:meta_title, "#{@pagetitle}")
```
Comment (dmarcoux) [inline src/api/app/views/webui/user/show.html.haml]: No need for interpolation

File: src/api/app/views/webui2/webui/user/show.html.haml [hunk 6]
```diff
@@ -1,4 +1,8 @@
-- @pagetitle = "Home of #{@displayed_user}"
+:ruby
+  @pagetitle = "Home of #{@displayed_user}"
+
+  content_for(:meta_title, "#{@pagetitle}")
```
Comment (dmarcoux) [inline src/api/app/views/webui2/webui/user/show.html.haml]: No need for interpolation

File: src/api/app/views/layouts/webui/webui.html.erb [hunk 7]
```diff
@@ -9,6 +9,13 @@
     <%= favicon_link_tag %>
 
     <title><%= "#{@pagetitle} - " if @pagetitle -%><%= home_title %></title>
+    <meta property="og:title" content="<%= meta_title %>"/>
+    <meta property="og:type" content="website"/>
+    <meta property="og:url" content="<%= request.original_url %>"/>
+    <meta property="og:image" content="<%= meta_image %>"/>
+    <% if meta_description %>
+    <meta property="og:description" content="<%= meta_description %>"/>
```
Comment (dmarcoux) [inline src/api/app/views/layouts/webui/webui.html.erb]: This line should be indented with 2 spaces.

File: src/api/app/helpers/meta_tag_helper.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+module MetaTagHelper
+  def meta_title
+    content_for?(:meta_title) ? content_for(:meta_title) : 'Open Build Service'
```
Comment (dmarcoux) [inline src/api/app/helpers/meta_tag_helper.rb]: TIL, I didn't know `content_for?` existed. I looked at the source code to see what it does exactly.

So [`content_for` does this](https://github.com/rails/rails/blob/e69ff43060c1194d2a3bd9b8d9e23f3ae26b84b5/actionview/lib/action_view/helpers/capture_helper.rb#L162) when you only pass it a name:
```
@view_flow.get(name).presence
```

[`content_for?` checks](https://github.com/rails/rails/blob/e69ff43060c1194d2a3bd9b8d9e23f3ae26b84b5/actionview/lib/action_view/helpers/capture_helper.rb#L192) if it's present with:
```
@view_flow.get(name).present?
```

With this in mind, we can change this method's code to
```
content_for(:meta_title) || 'Open Build Service'
```

Same for `meta_image` below.

**Comment by bgeuken** [comment]
@hellcp @hennevogel *ping* Since you were discussing this already in the issue... In case you want to review this one

File: src/api/app/helpers/meta_tag_helper.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  end
+
+  def meta_image
+    content_for?(:meta_image) ? content_for(:meta_image) : image_url('obs-logo_small.svg')
```
Comment (hellcp) [inline src/api/app/helpers/meta_tag_helper.rb]: SVG images will not be supported as display by many providers, due to possibility of executing code within SVG (via embedded js), use PNG instead (you might need to render SVG to PNG in assets dir).

File: src/api/app/views/layouts/webui2/webui.html.haml (hunk 1 above)
Comment (hellcp) [reply src/api/app/views/layouts/webui2/webui.html.haml]: Additionally to the title I would add `og:site_name`, with instance name.

Comment (SDBowen) [reply src/api/app/views/layouts/webui2/webui.html.haml]: Added `og:site_name` and removed all the extra interpolation!

File: src/api/app/views/layouts/webui2/webui.html.haml [hunk 10]
```diff
@@ -9,6 +9,14 @@
     %title
       = "#{@pagetitle} - " if @pagetitle
       = home_title
+    %meta{ property: "og:title", content: meta_title}/
```
Comment (coolo) [inline src/api/app/views/layouts/webui2/webui.html.haml]: these slashes at the end look too much

File: src/api/app/views/webui/request/show.html.erb (hunk 2 above)
Comment (SDBowen) [reply src/api/app/views/webui/request/show.html.erb]: I originally assumed that a bigger image would be best due to wherever these links get shared, the image will be resized to the site/app's open graph styling.

I tried to test this after reading your comment and found the picture wasn't loading at all, Oops! I ended up not using `gravatar_icon` and using the default Gravatar image size.

I will write some more about this below.

File: src/api/app/views/layouts/webui2/webui.html.haml (hunk 3 above)
Comment (SDBowen) [reply src/api/app/views/layouts/webui2/webui.html.haml]: I guess I went a little over board on the interpolation! Cleaned all this up.

File: src/api/app/views/layouts/webui/webui.html.erb (hunk 7 above)
Comment (SDBowen) [reply src/api/app/views/layouts/webui/webui.html.erb]: Done!

File: src/api/app/helpers/meta_tag_helper.rb (hunk 8 above)
Comment (SDBowen) [reply src/api/app/helpers/meta_tag_helper.rb]: Nice, this made things much cleaner!

File: src/api/app/helpers/meta_tag_helper.rb (hunk 9 above)
Comment (SDBowen) [reply src/api/app/helpers/meta_tag_helper.rb]: Good catch. I made a PNG and used that instead.

**Comment by SDBowen** [comment]
Attempting to test these in slack, I ran into an issue.

I was originally using the `gravatar_icon` helper to return the URL of a user's image, but found that the addition of the default image parameter `d=wavatar` prevented the og:image tag from getting pickup up. The returned URL did not end in a file-type extension.

Gravatar allows you to supply your own default image, which can give the URL a PNG ending. Adding a default image solved most the issues.

Everything seems to be working fine, except the `/request/show`. I could not get the open graph data to show. 

Some testing in Slack: (All but the Request page worked)

Screenshot from 2019-03-18 23-58-36

**Comment by coolo** [comment]
https://opengraphcheck.com/result.php?url=https%3A%2F%2Fobs-reviewlab.opensuse.org%2Fsdbowen-add-meta-tags%2Frequest%2Fshow%2F2#.XJCGdqyYWHt looks good though

File: src/api/app/views/layouts/webui2/webui.html.haml (hunk 10 above)
Comment (coolo) [reply src/api/app/views/layouts/webui2/webui.html.haml]: Just as I thought:
bs/src/api/app/views/layouts/webui2/webui.html.haml:13: syntax error, unexpected tIDENTIFIER, expecting ')'

**Comment by coolo** [comment]
https://chrome.google.com/webstore/detail/open-graph-preview/ehaigphokkgebnmdiicabhjhddkaekgh?hl=de shows the infos as well. Rocket chat seems to ignore them completely - but in all routes.

File: src/api/app/views/layouts/webui2/webui.html.haml (hunk 10 above)
Comment (SDBowen) [reply src/api/app/views/layouts/webui2/webui.html.haml]: Oh no... I missed a comma.

Should all these slashes be removed? I see in the HAML docs that meta tags are treated as being empty. So I guess they are a bit redundant.

**Comment by coolo** [comment]
Looking at https://github.com/RocketChat/Rocket.Chat/blob/ed092fbe490c21d64c071772ce1da66515837353/app/oembed/client/oembedUrlWidget.js it doesn't make sense it's ignored :(

File: src/api/app/views/layouts/webui2/webui.html.haml (hunk 10 above)
Comment (coolo) [reply src/api/app/views/layouts/webui2/webui.html.haml]: They are empty by default - so just remove them.

**Comment by coolo** [comment]
I still don't understand why rocketchat ignores them though

File: src/api/app/views/webui2/webui/package/show.html.haml [hunk 11]
```diff
@@ -2,6 +2,9 @@
   @pagetitle = "Show #{@project} / #{truncate(@package.name, length: 50)}"
   devel_package = @package.develpackage
 
+  content_for(:meta_title, @package.title.presence || @package.name)
```
Comment (eduardoj) [inline src/api/app/views/webui2/webui/package/show.html.haml]: [Nitpick]: `@package.name` would also be written as `@package`, as it already happens with `@project.name`. The same applies to `src/api/app/views/webui/package/show.html.erb`.

**Comment by dmarcoux** [comment]
Great work!

**Comment by hennevogel** [comment]
Awesome :bouquet:

**Pull request merged**